### PR TITLE
Only resume uploads to the same location as the user requested (fixes #139)

### LINKED ISSUE #139: Re-uploading an existing file to a new AWS key tries to resume an upload to the original key
So I have a use case where a user might want to upload the same file to different buckets/keys, but right now if the same file is selected, then it fetches it out of localStorage and tries to resume the upload to an existing key. I'd like to alter the logic so it'll only resume the upload if the key hasn't changed (I believe it needs to be done around here https://github.com/TTLabs/EvaporateJS/blob/master/evaporate.js#L881)

I'll raise a PR for this tomorrow unless anyone objects?

Thanks

### PR DESCRIPTION
**Pull request by cnorthwood**

### CODE REVIEW

**Comment by bikeath1337** [comment]
Thanks...been a bit busy, we review later. Thank you, @cnorthwood

File: evaporate.js [hunk 1]
```diff
@@ -885,7 +886,7 @@
                 }
                 var completedAt = new Date(u.completedAt || FAR_FUTURE);
 
-                return con.partSize === u.partSize && completedAt > HOURS_AGO;
+                return con.partSize === u.partSize && completedAt > HOURS_AGO && me.name === u.awsKey && con.bucket === u.bucket;
```
Comment (bikeath1337) [inline evaporate.js]: @cnorthwood This looks good but the `me.name === uwsKey` breaks existing functionality. Checking for the bucket is definitely something that needs to be added, though.

I think that EvaporateJS already gives you what you need in it's config options `allowS3ExistenceOptimization` (defaults to `false`).

Comment (cnorthwood) [inline evaporate.js]: I'm still not convinced that Evaporate uploading to a different location to the one you ask for is ever the right thing to do - I'd argue that the current behaviour is a bug.

Comment (cnorthwood) [inline evaporate.js]: If you've uploaded the same file to a different location previously, even when `allowS3ExistenceOptimization` is true, it still generates requests to the AWS key of the first time you uploaded the file, not the second - then instantly completes the upload (but to the old key)

Comment (bikeath1337) [inline evaporate.js]: I think we need to make sure we're describing the same thing. We're using what you describe to be a bug -- at the rate of 100K + uploads per month. EvaporateJS does not upload to a different location -- and by that I mean, bucket. Given that the "right" way to be performant with S3 is to create randomized file names, the implemented feature uses heuristics (from the uploader's perspective) to save uploading costs for large files. The frequency with which uploads fail for batch file uploading of big files is surprisingly large, thus the optimization.

You only get the optimization if you enable it explicitly. It's disabled by default.

If you still want to use the feature with your restrictions, then I'm happy to consider an enhancement that enriches or restricts the existing functionality without breaking usage. For example, in your case, you want to say, "hey, only optimize if the uploaded object name === S3 object name". That's definitely something that would be an enhancement, disabled by default, and enabled by you in your use cases.

Comment (cnorthwood) [inline evaporate.js]: by location, I mean randomly generated AWS key. My workflow works like this:

* Client asks server for a randomly generated key, associated with a particular user
* Server returns key to client, and marks that that user is the author of that key in the database and the upload is "in progress"
* Client then uses Evaporate to upload that file
* Server checks that the key that is being asked to signed is one that user is authorised to upload to
* Once the upload completes, the client notifies the server, and the server marks the upload as completed in the database, and runs the appropriate process (in our case, it's a video transcode - as an aside, these origin files are deleted once the transcode is complete and the transcode is saved to another bucket)

When the user tries to upload that file a second time, we saw Evaporate sending a completely different AWS key to be signed. As we don't let users overwrite files, the server refused to sign the request. It took a lot of head scratching to figure out why Evaporate wasn't paying attention to the `aws_key` parameter we were passing in, and instead trying to upload it to a different location.

I think that's a more normal workflow, but we can agree to disagree. I'll modify the PR to add an option and a warning.

**Comment by cnorthwood** [comment]
sorry, just updated mine above, I was using key in the context of the AWS docs describes an S3 key, as opposed to an Access Key or similar (I've realised the Evaporate docs refer to this as the file name)

File: evaporate.js (hunk 1 above)
Comment (cnorthwood) [inline evaporate.js]: just updated my above comment - I've been using key in the context of the S3 docs, which Evaporate refers to as the file name, as opposed to the Access Key

**Comment by cnorthwood** [comment]
PR updated @bikeath1337

**Comment by bikeath1337** [comment]
Thanks, @cnorthwood

**Comment by bikeath1337** [comment]
Thank you. Will package this up a bit later in the week.

**Pull request merged**